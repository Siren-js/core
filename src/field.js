import * as coerce from './util/coerce';
import { isNullish, isString } from './util/type-guard';

export class Field {
    #name;
    #class;
    #title;
    #type;
    #value;

    constructor(name, options = {}) {
        if (!isString(name)) {
            throw new TypeError('Field.name must be a string');
        }

        const { class: fieldClass, title, type, value, ...extensions } =
            options ?? {};

        this.#name = name;
        this.class = fieldClass;
        this.title = title;
        this.type = type;
        this.value = value;

        Object.keys(extensions).forEach((key) => {
            this[key] = extensions[key];
        });
    }

    get name() {
        return this.#name;
    }

    get class() {
        return this.#class;
    }

    set class(value) {
        this.#class = coerce.toOptionalStringArray(value, this.class);
    }

    get title() {
        return this.#title;
    }

    set title(value) {
        this.#title = coerce.toOptionalString(value, this.title);
    }

    get type() {
        return this.#type;
    }

    set type(value) {
        this.#type = coerce.toOptionalString(value, this.type);
    }

    get value() {
        return this.#value;
    }

    set value(value) {
        this.#value = isNullish(value) ? undefined : value;
    }
}
