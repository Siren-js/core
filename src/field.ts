/**
 * Copyright (C) 2020  Dillon Redding
 *
 * See the full notice in the LICENSE file at the top level of the repository.
 */
import * as coerce from './coerce';
import deepFreeze from './deep-freeze';
import Extendable from './extendable';
import { isRecord, isString, isStringArray, isUndefined } from './type-guards';

/**
 * Creates a `ParsedField` object
 * @param name Name of the field
 * @param options Object containing optional field members (e.g., `title`, `type`) and extensions
 */
export function field<T>(name: string, options: OptionalFieldMembers<T> = {}): ParsedField<T> {
    const { 'class': fieldClass, title, type, value, ...extensions } = options;
    return deepFreeze({
        name: coerce.toString(name),
        class: coerce.toOptionalStringArray(fieldClass),
        title: coerce.toOptionalString(title),
        type: coerce.toOptionalString(type),
        value,
        ...extensions,

        update<U>(value: U): ParsedField<U> {
            return field(name, { ...options, value });
        }
    });
}

export type OptionalFieldMembers<T> = Extendable & Pick<ParsableField<T>, 'class' | 'title' | 'type' | 'value'>;

export interface ParsableField<T = unknown> extends Extendable {
    class?: string[] | string;
    name: string;
    title?: string;
    type?: string;
    value?: T;
}

export interface ParsedField<T = unknown> extends Field<T> {
    /**
     * Update the field's value.
     * @returns A new `Field` object with the updated value
     */
    update<U>(value: U): ParsedField<U>;
}

/**
 * Describes a Siren [field](https://github.com/kevinswiber/siren#fields-1),
 * which represents a control inside of an `Action`.
 */
export interface Field<T = unknown> extends Extendable {
    /**
     * Describes aspects of the field based on the current representation.
     */
    readonly class?: readonly string[];

    /**
     * A name describing the control.
     */
    readonly name: string;

    /**
     * Textual annotation of a field. Clients may use this as a label.
     */
    readonly title?: string;

    /**
     * The input type of the field. This may include
     * [input types](https://html.spec.whatwg.org/#the-input-element) specified
     * in HTML5 (see the `FieldType` enumeration). The default value is
     * `'text'`.
     */
    readonly type?: string;

    /**
     * A value assigned to the field.
     */
    readonly value?: Readonly<T>;
}

/**
 * Determines whether `value` is a valid Siren field. Note that this is a strict
 * check. To see if `value` is parsable by `field()`, use `isParsableField()`.
 */
export function isField(value: unknown): value is Field {
    return isParsableField(value) &&
        (isUndefined(value.class) || isStringArray(value.class)) &&
        (isUndefined(value.title) || isString(value.title)) &&
        (isUndefined(value.type) || isString(value.type));
}

/**
 * Determines whether `value` is a parsable Siren field. This function differs
 * from `isField()` in that the loose parsing rules of `field()` are taken into
 * consideration.
 */
export function isParsableField(value: unknown): value is ParsableField {
    return isRecord(value) && isString(value.name);
}

/**
 * Defines [input types](https://html.spec.whatwg.org/#attr-input-type-keywords)
 * specified in HTML5. `enum` names correspond to the input state and the values
 * correspond to the input keyword.
 */
export enum FieldType {
    Hidden = 'hidden',
    Text = 'text',
    Search = 'search',
    Telephone = 'tel',
    Url = 'url',
    Email = 'email',
    Password = 'password',
    Date = 'date',
    Month = 'month',
    Week = 'week',
    Time = 'time',
    LocalDateAndTime = 'datetime-local',
    Number = 'number',
    Range = 'range',
    Color = 'color',
    Checkbox = 'checkbox',
    RadioButton = 'radio',
    FileUpload = 'file',
    SubmitButton = 'submit',
    ImageButton = 'image',
    ResetButton = 'reset',
    Button = 'button'
}
