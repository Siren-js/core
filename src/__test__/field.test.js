import { Field } from '../field';

const name = 'quantity';

describe('Field constructor', () => {
    it('should throw TypeError when given no arguments', () => {
        expect(() => new Field()).toThrow(TypeError);
    });

    it('should handle null options gracefully', () => {
        expect(() => new Field(name, null)).not.toThrow();
    });
});

describe('Field.name', () => {
    it('should accept any string', () => {
        [name, 'order number', ''].forEach((value) => {
            const field = new Field(value);
            expect(field.name).toEqual(value);
        });
    });

    it('should throw error given non-string in constructor', () => {
        [undefined, null, true, 42, [true, 42, 'foo'], { foo: 'bar' }].forEach(
            (value) => {
                expect(() => new Field(value)).toThrow(TypeError);
            }
        );
    });

    it('should be read-only', () => {
        const field = new Field(name);
        expect(() => (field.name = 'foo')).toThrow(TypeError);
    });
});

describe('Field.class', () => {
    it('should accept any array of strings', () => {
        [['integer'], ['positive', 'integer'], []].forEach((value) => {
            let field = new Field(name, { class: value });
            expect(field.class).toEqual(value);
        });
    });

    it('should coerce string to singleton array', () => {
        ['integer', ''].forEach((value) => {
            const field = new Field(name, { class: value });
            expect(field.class).toEqual([value]);
        });
    });

    it('should allow undefined and coerce null', () => {
        [undefined, null].forEach((value) => {
            const field = new Field(name, { class: value });
            expect(field.class).toBeUndefined();
        });
    });

    it('should remove non-strings from array', () => {
        const field = new Field(name, {
            class: [true, 42, 'integer', null, undefined]
        });

        expect(field.class).toEqual(['integer']);
    });

    it('should ignore invalid value', () => {
        [true, 42, {}].forEach((value) => {
            const field = new Field(name, { class: value });
            expect(field.class).toBeUndefined();
        });
    });
});

describe('Field.title', () => {
    const title = 'This is a Field';

    it('should accept any string', () => {
        [title, 'Å Fâñçÿ Títlè', ''].forEach((value) => {
            let field = new Field(name, { title: value });
            expect(field.title).toEqual(value);

            field = new Field(name);
            field.title = value;
            expect(field.title).toEqual(value);
        });
    });

    it('should ignore non-string value', () => {
        [true, 42, [], {}].forEach((value) => {
            let field = new Field(name, { title: value });
            expect(field.title).toBeUndefined();

            field = new Field(name, { title });
            field.title = value;
            expect(field.title).toEqual(title);
        });
    });

    it('should allow undefined and coerce null to undefined', () => {
        [undefined, null].forEach((value) => {
            let field = new Field(name, { title: value });
            expect(field.title).toBeUndefined();

            field = new Field(name, { title });
            field.title = value;
            expect(field.title).toBeUndefined();
        });
    });
});

describe('Field.type', () => {
    const type = 'number';

    it('should accept any string', () => {
        [type, 'nümbér', ''].forEach((value) => {
            let field = new Field(name, { type: value });
            expect(field.type).toEqual(value);

            field = new Field(name);
            field.type = value;
            expect(field.type).toEqual(value);
        });
    });

    it('should ignore non-string value', () => {
        [true, 42, [], {}].forEach((value) => {
            let field = new Field(name, { type: value });
            expect(field.type).toBeUndefined();

            field = new Field(name, { type });
            field.type = value;
            expect(field.type).toEqual(type);
        });
    });

    it('should allow undefined and coerce null to undefined', () => {
        [undefined, null].forEach((value) => {
            let field = new Field(name, { type: value });
            expect(field.type).toBeUndefined();

            field = new Field(name, { type });
            field.type = value;
            expect(field.type).toBeUndefined();
        });
    });
});

describe('Field.value', () => {
    it('should accept any non-nullish value', () => {
        [true, 42, 'foo', [true, 42, 'foo'], { foo: 'bar' }].forEach(
            (value) => {
                const field = new Field(name, { value });
                expect(field.value).toEqual(value);
            }
        );
    });

    it('should allow undefined and coerce null to undefined', () => {
        [undefined, null].forEach((value) => {
            let field = new Field(name, { value });
            expect(field.value).toBeUndefined();

            field = new Field(name, { value: 42 });
            field.value = value;
            expect(field.value).toBeUndefined();
        });
    });
});

test('Field extensions', () => {
    const min = 1;
    const step = '2';

    const field = new Field(name, { min });
    field.step = step;

    expect(field.min).toEqual(min);
    expect(field.step).toEqual(step);
});

test('Field serialization', () => {
    const field = new Field(name, {
        class: ['integer'],
        title: 'Quantity',
        type: 'number',
        value: 9,
        min: 1,
        step: '2'
    });

    const json = JSON.stringify(field, null, 2);

    expect(json).toMatchSnapshot();
});
