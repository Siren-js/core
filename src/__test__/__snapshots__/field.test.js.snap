// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field serialization 1`] = `
"{
  \\"min\\": 1,
  \\"step\\": \\"2\\"
}"
`;
